lean_source/06_Abstract_Algebra/source_01_Structures.lean

```diff
@@ -0,0 +1,503 @@
+An *instance* of the structure is a particular bundle of data satisfying
+the constraints. For example, we can specify that a point is
+a tuple of three real numbers:
+BOTH: -/
+-- QUOTE:
+@[ext] structure point := (x : ℝ) (y : ℝ) (z : ℝ)
+-- QUOTE.
+
+/- TEXT:
+The ``@[ext]`` annotation tells Lean to automatically generate theorems
+that can be used to prove that two instances of a structure are equal
+when their components are equal.
+EXAMPLES: -/
+-- QUOTE:
+#check point.ext
+
+example (a b : point) (hx : a.x = b.x) (hy : a.y = b.y) (hz : a.z = b.z) :
+  a = b :=
+begin
+  ext,
+  repeat { assumption }
+end
+-- QUOTE.
+
+/- TEXT:
+We can then define particular instances of the ``point`` structure.
+Lean provides multiple ways of doing that.
+EXAMPLES: -/
+-- QUOTE:
+def my_point1 : point :=
+{ x := 2,
+  y := -1,
+  z := 4 }
+
+def my_point2 :=
+{ point .
+  x := 2,
+  y := -1,
+  z := 4 }
+
+def my_point3 : point := ⟨2, -1, 4⟩
+
+def my_point4 := point.mk 2 (-1) 4
+-- QUOTE.
+
+/- TEXT:
+In the first two examples, the fields of the structure are named
+explicitly.
+In the first case, because Lean knows that the expected type of
+``my_point1`` is a ``point``, you can start the definition by
+writing ``{! !}``. Clicking on the light bulb
+that appears nearby in VS Code will then
+give you the option of inserting a template definition
+with the field names listed for you.
+
+By default, the *constructor* for the ``point`` structure is named
+``point.mk``. You can specify a different name if you want.
+EXAMPLES: -/
+-- QUOTE:
+structure point' := build :: (x : ℝ) (y : ℝ) (z : ℝ)
+
+#check point'.build 2 (-1) 4
+-- QUOTE.
+
+/- TEXT:
+The next two examples show how to define functions
+on structures using pattern matching.
+Whereas the second example makes the ``point.mk``
+constructor explicit, the first example uses anonymous constructors
+for brevity.
+Lean can infer the relevant constructor from the indicated type of
+``add``.
+It is conventional to put definitions and theorems associated
+with a structure like ``point`` in a namespace with the same name.
+In the example below, the full name of ``add`` is ``point.add``,
+which allows us to use the anonymous projection notation.
+BOTH: -/
+-- QUOTE:
+namespace point
+
+def add : point → point → point
+| ⟨x₁, y₁, z₁⟩ ⟨x₂, y₂, z₂⟩ := ⟨x₁ + x₂, y₁ + y₂, z₁ + z₂⟩
+
+-- EXAMPLES:
+def add' : point → point → point
+| (point.mk x₁ y₁ z₁) (point.mk x₂ y₂ z₂) :=
+  { x := x₁ + x₂,
+    y := y₁ + y₂,
+    z := z₁ + z₂ }
+
+#check add my_point1 my_point2
+#check my_point1.add my_point2
+
+end point
+
+#check point.add my_point1 my_point2
+#check my_point1.add my_point2
+-- QUOTE.
+
+/- TEXT:
+Below we will continue to put definitions in the relevant
+namespace, but we will leave the namespacing commands out of the quoted
+snippets.
+
+Within a proof, structures can be decomposed with the ``cases`` or ``rcases``
+tactics. You can also use ``rintros`` or a pattern-matching
+lambda.
```

Comment:
It would be really nice to have a way to clearly distinguish sentences that introduce new tactics from references to already covered tactics. This could be either a textual convention or some kind of cross linking. We should also consider whether such a new use of tactics is worth a mention in the index.